paddle/fluid/operators/mkldnn/test_mkldnn_caching.cc

```diff
@@ -0,0 +1,188 @@
+                 const framework::DDim &dims, const std::string &output_name,
+                 bool inplace = false) {
+  framework::Scope scope;
+
+  std::map<const std::string, int> num_inputs = {
+      {"softmax", 1}, {"relu", 1}, {"elementwise_add", 2}};
+
+  std::string first_input = inplace == true ? output_name : "x";
+
+  std::vector<InputVars> input_names = {
+      {first_input, scope.Var(first_input)->GetMutable<framework::LoDTensor>()},
+      {"x1", num_inputs[op_type] > 1
+                 ? scope.Var("x1")->GetMutable<framework::LoDTensor>()
+                 : nullptr},
+      {"x2", num_inputs[op_type] > 2
+                 ? scope.Var("x2")->GetMutable<framework::LoDTensor>()
+                 : nullptr},
+      {"x3", num_inputs[op_type] > 3
+                 ? scope.Var("x3")->GetMutable<framework::LoDTensor>()
+                 : nullptr},
+      {"x4", num_inputs[op_type] > 4
+                 ? scope.Var("x4")->GetMutable<framework::LoDTensor>()
+                 : nullptr}};
+  auto *y = scope.Var(output_name)->GetMutable<framework::LoDTensor>();
+
+  // Initialize input data
+  std::uniform_real_distribution<T> dist(static_cast<T>(10.0),
+                                         static_cast<T>(20.0));
+  std::mt19937 engine;
+  size_t numel = static_cast<size_t>(framework::product(dims));
+  for (int i = 0; i < num_inputs[op_type]; ++i) {
+    input_names[i].tensor->Resize(dims);
+    auto data_ptr = input_names[i].tensor->mutable_data<T>(place);
+    for (size_t i = 0; i < numel; ++i) {
+      data_ptr[i] = dist(engine);
+    }
+  }
+
+  // Initialize output
+  y->Resize(dims);
+  auto y_ptr = y->mutable_data<T>(place);
+  for (size_t i = 0; i < numel; ++i) {
+    y_ptr[i] = static_cast<T>(0);
+  }
+
+  auto &pool = platform::DeviceContextPool::Instance();
+
+  // In-place computation
+  if (inplace) {
+    auto op = num_inputs[op_type] > 1
+                  ? framework::OpRegistry::CreateOp(
+                        op_type, {{"X", {first_input}}, {"Y", {"x1"}}},
+                        {{"Out", {output_name}}}, {{"use_mkldnn", {true}}})
+                  : framework::OpRegistry::CreateOp(
+                        op_type, {{"X", {first_input}}},
+                        {{"Out", {output_name}}}, {{"use_mkldnn", {true}}});
+
+    op->Run(scope, place);
+    pool.Get(place)->Wait();
+  } else {
+    // Out of place computation
+    auto op_ref = num_inputs[op_type] > 1
+                      ? framework::OpRegistry::CreateOp(
+                            op_type, {{"X", {first_input}}, {"Y", {"x1"}}},
+                            {{"Out", {output_name}}}, {{"use_mkldnn", {true}}})
+                      : framework::OpRegistry::CreateOp(
+                            op_type, {{"X", {first_input}}},
+                            {{"Out", {output_name}}}, {{"use_mkldnn", {true}}});
+
+    op_ref->Run(scope, place);
+    pool.Get(place)->Wait();
```

Comment:
Those two `if` branches contains exactly the same code. Are they necessary?

Reply:
good catch! thanks